### PR TITLE
pcre2: fix invalid pcre2-config output for static libs

### PR DESCRIPTION
**Pull request by valgur**
`pcre2` includes a `pcre2-config` executable that outputs necessary compiler and linker flags to link against `pcre2`. However, it currently outputs invalid library names for static versions of the libraries.

This is a blocker for #19058, as this causes linking against `pcre2` to fail.

### CODE REVIEW

**Comment by uilianries** [comment]
I just launched an internal CI build to generate pcre2/10.32

**Review by jwillikers — APPROVED** [review]

**Comment by ptc-rgrasset** [comment]
Any update on this CR?

We are waiting on it for some of our dependencies.

**Review by danimtb — CHANGES_REQUESTED** [review]
this fix looks ok but I think there is no need to add that complexity to the test_package. Please remove the test_package changes

File: recipes/pcre2/all/test_package/conanfile.py [hunk 1]
```diff
@@ -1,26 +1,56 @@
 from conan import ConanFile
 from conan.tools.build import can_run
+from conan.tools.files import save, load
 from conan.tools.cmake import CMake, cmake_layout
+from conan.tools.scm import Version
```
Comment (danimtb) [inline recipes/pcre2/all/test_package/conanfile.py:5]:
```suggestion
```

File: recipes/pcre2/all/test_package/conanfile.py [hunk 2]
```diff
@@ -1,26 +1,56 @@
...
 import os
+from io import StringIO
```
Comment (danimtb) [inline recipes/pcre2/all/test_package/conanfile.py:7]:
```suggestion
```

File: recipes/pcre2/all/test_package/conanfile.py [hunk 3]
```diff
@@ -1,26 +1,56 @@
...
 
 class TestPackageConan(ConanFile):
     settings = "os", "arch", "compiler", "build_type"
     generators = "CMakeToolchain", "CMakeDeps", "VirtualRunEnv"
     test_type = "explicit"
 
+    @property
+    def _settings_build(self):
+        return getattr(self, "settings_build", self.settings)
+
```
Comment (danimtb) [inline recipes/pcre2/all/test_package/conanfile.py:18]:
```suggestion
```

File: recipes/pcre2/all/test_package/conanfile.py [hunk 4]
```diff
@@ -1,26 +1,56 @@
...
+        if self._settings_build.os == "Windows":
+            if not self.conf.get("tools.microsoft.bash:path", check_type=str):
+                self.tool_requires("msys2/cci.latest")
+
+    def generate(self):
+        # Workaround for Conan v1: store dependency info for later use in test()
+        pcre2_cpp_info = self.dependencies["pcre2"].cpp_info.aggregated_components()
+        save(self, os.path.join(self.build_folder, "bindir"), pcre2_cpp_info.bindir)
+        save(self, os.path.join(self.build_folder, "libs"), " ".join(pcre2_cpp_info.libs))
+
```
Comment (danimtb) [inline recipes/pcre2/all/test_package/conanfile.py:35]:
```suggestion
```

File: recipes/pcre2/all/test_package/conanfile.py [hunk 5]
```diff
@@ -1,26 +1,56 @@
...
+            if Version(self.tested_reference_str.split("/")[1]) >= "10.38":
+                bindir = load(self, os.path.join(self.build_folder, "bindir"))
+                libs = load(self, os.path.join(self.build_folder, "libs")).split(" ")
+                output = StringIO()
+                self.run(f"bash {bindir}/pcre2-config --libs8", output)
+                ldflags_str = next(l for l in output.getvalue().splitlines() if l.lower().startswith("-l")).strip()
+                ldflags = ldflags_str.split()
+                for flag in ldflags:
+                    if flag.startswith("-l"):
+                        assert flag[2:] in libs, f"Invalid library target '{flag[2:]}' output by pcre2-config. Not found in {libs}."
```
Comment (danimtb) [inline recipes/pcre2/all/test_package/conanfile.py:56]:
```suggestion
```

**Review by danimtb — COMMENTED** [review]
no need to add that complexity to test_package

**Review by danimtb — APPROVED** [review]

**Review by jwillikers — APPROVED** [review]

**Pull request merged**